File: code_puppy/config/paths.py
"""XDG-compatible path resolution for Code Puppy directories and files.

Mirrors ``CodePuppyControl.Config.Paths`` in the Elixir runtime.

ADR-003: All path resolution respects pup-ex isolation. When running as
pup-ex (``PUP_EX_HOME`` set), paths resolve under ``~/.code_puppy_ex/``
and NEVER under ``~/.code_puppy/``.

This module provides lazy path accessors for INTERNAL use (``_xdg_*()``,
``_path_*()``) and a ``__getattr__``-based ``_LAZY_PATH_FACTORIES`` dict
for backward-compatible EXTERNAL access (e.g. ``config.CONFIG_FILE``).
"""

# ruff: noqa: F822 — lazy __all__ names resolved via __getattr__ (PEP 562)

from __future__ import annotations

import os
import pathlib
from collections.abc import Callable

# Re-export the core config_paths functions for convenience
from code_puppy.config_paths import (
    ConfigIsolationViolation,
    resolve_path,
    safe_append,
    safe_atomic_write,
    safe_mkdir_p,
    safe_rm,
    safe_rm_rf,
    safe_write,
    with_sandbox,
)
from code_puppy.config_paths import (
    cache_dir as _cp_cache_dir,
)
from code_puppy.config_paths import (
    config_dir as _cp_config_dir,
)
from code_puppy.config_paths import (
    data_dir as _cp_data_dir,
)
from code_puppy.config_paths import (
    state_dir as _cp_state_dir,
)

__all__ = [
    # Path constants (lazy via __getattr__)
    "STATE_DIR",
    "CONFIG_DIR",
    "CACHE_DIR",
    "AUTOSAVE_DIR",
    "EXTRA_MODELS_FILE",
    # Re-exports from config_paths
    "ConfigIsolationViolation",
    "safe_write",
    "safe_mkdir_p",
    "safe_rm",
    "safe_rm_rf",
    "safe_atomic_write",
    "safe_append",
    "with_sandbox",
    "resolve_path",
]


# ---------------------------------------------------------------------------
# XDG dir helpers — INTERNAL use only
# ---------------------------------------------------------------------------


def _get_xdg_dir(env_var: str, fallback: str) -> str:
    """Get XDG directory respecting pup-ex isolation.

    Delegates to config_paths for the canonical implementation.
    """
    from code_puppy.config.loader import _get_xdg_dir as _loader_xdg

    return _loader_xdg(env_var, fallback)


def _xdg_config_dir() -> str:
    """Return the XDG config directory (respects pup-ex isolation)."""
    return _cp_config_dir()


def _xdg_data_dir() -> str:
    """Return the XDG data directory (respects pup-ex isolation)."""
    return _cp_data_dir()


def _xdg_cache_dir() -> str:
    """Return the XDG cache directory (respects pup-ex isolation)."""
    return _cp_cache_dir()


def _xdg_state_dir() -> str:
    """Return the XDG state directory (respects pup-ex isolation)."""
    return _cp_state_dir()


# ---------------------------------------------------------------------------
# Path accessors — INTERNAL use only
# ---------------------------------------------------------------------------


def _override_str(name: str) -> str | None:
    """Check if a module-level path constant has been overridden."""
    # This is kept for backward compat with test monkeypatching.
    # Since we no longer have module-level constants, always returns None
    # unless a test injects via _LAZY_PATH_OVERRIDES.
    return _LAZY_PATH_OVERRIDES.get(name)


def _override_path(name: str) -> pathlib.Path | None:
    """Check if a module-level path constant has been overridden (Path)."""
    val = _LAZY_PATH_OVERRIDES.get(name)
    if val is None:
        return None
    return pathlib.Path(val)


# Test override dict — tests can set keys here to override paths
_LAZY_PATH_OVERRIDES: dict[str, str] = {}


def _path_config_file() -> pathlib.Path:
    """Return the config file path (respects pup-ex isolation)."""
    override = _override_path("CONFIG_FILE")
    if override is not None:
        return override
    return pathlib.Path(_xdg_config_dir()) / "puppy.cfg"


def _path_mcp_servers_file() -> pathlib.Path:
    """Return the MCP servers file path (respects pup-ex isolation)."""
    override = _override_path("MCP_SERVERS_FILE")
    if override is not None:
        return override
    return pathlib.Path(_xdg_config_dir()) / "mcp_servers.json"


def _path_agents_dir() -> pathlib.Path:
    """Return the agents directory path (respects pup-ex isolation)."""
    override = _override_path("AGENTS_DIR")
    if override is not None:
        return override
    return pathlib.Path(_xdg_data_dir()) / "agents"


def _path_skills_dir() -> pathlib.Path:
    """Return the skills directory path (respects pup-ex isolation)."""
    override = _override_path("SKILLS_DIR")
    if override is not None:
        return override
    return pathlib.Path(_xdg_data_dir()) / "skills"


def _path_autosave_dir() -> pathlib.Path:
    """Return the autosave directory path (respects pup-ex isolation)."""
    override = _override_path("AUTOSAVE_DIR")
    if override is not None:
        return override
    return pathlib.Path(_xdg_cache_dir()) / "autosaves"


def _path_command_history_file() -> pathlib.Path:
    """Return the command history file path (respects pup-ex isolation)."""
    override = _override_path("COMMAND_HISTORY_FILE")
    if override is not None:
        return override
    return pathlib.Path(_xdg_state_dir()) / "command_history.txt"


def _path_default_sqlite_file() -> pathlib.Path:
    """Return the DBOS SQLite file path (respects pup-ex isolation)."""
    override = _override_path("_DEFAULT_SQLITE_FILE")
    if override is not None:
        return override
    return pathlib.Path(_xdg_data_dir()) / "dbos_store.sqlite"


# ---------------------------------------------------------------------------
# Lazy path factories for backward-compatible external access
# ---------------------------------------------------------------------------

_LAZY_PATH_FACTORIES: dict[str, Callable[[], object]] = {
    "CONFIG_DIR": lambda: _xdg_config_dir(),
    "DATA_DIR": lambda: _xdg_data_dir(),
    "CACHE_DIR": lambda: _xdg_cache_dir(),
    "STATE_DIR": lambda: _xdg_state_dir(),
    "CONFIG_FILE": lambda: _path_config_file(),
    "MCP_SERVERS_FILE": lambda: _path_mcp_servers_file(),
    "MODELS_FILE": lambda: pathlib.Path(_xdg_data_dir()) / "models.json",
    "EXTRA_MODELS_FILE": lambda: pathlib.Path(_xdg_data_dir()) / "extra_models.json",
    "AGENTS_DIR": lambda: _path_agents_dir(),
    "SKILLS_DIR": lambda: _path_skills_dir(),
    "CONTEXTS_DIR": lambda: pathlib.Path(_xdg_data_dir()) / "contexts",
    "_DEFAULT_SQLITE_FILE": lambda: _path_default_sqlite_file(),
    "CHATGPT_MODELS_FILE": lambda: (
        pathlib.Path(_xdg_data_dir()) / "chatgpt_models.json"
    ),
    "CLAUDE_MODELS_FILE": lambda: pathlib.Path(_xdg_data_dir()) / "claude_models.json",
    "AUTOSAVE_DIR": lambda: _path_autosave_dir(),
    "COMMAND_HISTORY_FILE": lambda: _path_command_history_file(),
    "DBOS_DATABASE_URL": lambda: os.environ.get(
        "DBOS_SYSTEM_DATABASE_URL",
        f"sqlite:///{_path_default_sqlite_file()}",
    ),
}


def __getattr__(name: str):
    """Lazy path resolution for external attribute access (PEP 562).

    External code doing ``from code_puppy.config import CONFIG_FILE``
    or ``config.CONFIG_FILE`` will hit this and get a freshly-computed
    value that respects pup-ex isolation.
    """
    if name in _LAZY_PATH_FACTORIES:
        return _LAZY_PATH_FACTORIES[name]()
    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")


# Clean up any pre-existing module-level names
for _lazy_name in _LAZY_PATH_FACTORIES:
    pass  # No-op; names are resolved via __getattr__

File: elixir/code_puppy_control/lib/code_puppy_control/config/paths.ex
defmodule CodePuppyControl.Config.Paths do
  @moduledoc """
  XDG-compatible path resolution for Code Puppy directories and files.

  Implements ADR-003 dual-home config isolation. The Elixir pup-ex runtime
  uses `~/.code_puppy_ex/` as its home, separate from the Python pup's
  `~/.code_puppy/`. Writes to the legacy home are blocked by the
  `CodePuppyControl.Config.Isolation` guard module.

  ## Home Resolution Priority

  1. `PUP_EX_HOME` env var — explicit Elixir home override
  2. `PUP_HOME` env var — legacy override (deprecation warning)
  3. `PUPPY_HOME` env var — oldest legacy override (deprecation warning)
  4. Default `~/.code_puppy_ex/` — standard fallback

  ## Path Layout

  ```
  ~/.code_puppy_ex/
  ├── puppy.cfg           # Main config (INI format)
  ├── mcp_servers.json    # MCP server definitions
  ├── models.json         # Model registry
  ├── extra_models.json   # User-added models
  ├── agents/             # Agent definitions
  ├── skills/             # Skill definitions
  ├── contexts/           # Context presets
  ├── autosaves/          # Session autosaves (cache)
  ├── command_history.txt # Command history (state)
  └── dbos_store.sqlite   # DBOS state store
  ```

  When XDG env vars are set, files split across config/data/cache/state
  directories per the spec. Otherwise everything lives under `~/.code_puppy_ex`.
  """

  require Logger

  @home_dir Path.expand("~")

  @max_symlink_depth 40

  # ── Base directories ────────────────────────────────────────────────────

  @doc """
  Root directory for all Elixir pup-ex files.

  Resolution order: `PUP_EX_HOME` → `PUP_HOME` (deprecated) →
  `PUPPY_HOME` (deprecated) → `~/.code_puppy_ex`.
  """
  @spec home_dir() :: String.t()
  def home_dir do
    cond do
      env = System.get_env("PUP_EX_HOME") ->
        env

      env = System.get_env("PUP_HOME") ->
        maybe_warn_deprecation("PUP_HOME")
        env

      env = System.get_env("PUPPY_HOME") ->
        maybe_warn_deprecation("PUPPY_HOME")
        env

      true ->
        Path.join(@home_dir, ".code_puppy_ex")
    end
  end

  @doc """
  Legacy home directory — always `~/.code_puppy`, regardless of env vars.

  This is the Python pup's home directory. Elixir pup-ex must NEVER write
  here. Use only for read-only import via `mix pup_ex.import`.

      read_only: true, legacy: true
  """
  @spec legacy_home_dir() :: String.t()
  def legacy_home_dir do
    Path.join(@home_dir, ".code_puppy")
  end

  @doc """
  Configuration directory. Contains `puppy.cfg`, `mcp_servers.json`.

  Uses `XDG_CONFIG_HOME/code_puppy_ex` if set, otherwise falls back to `home_dir/`.
  """
  @spec config_dir() :: String.t()
  def config_dir do
    xdg_dir("XDG_CONFIG_HOME")
  end

  @doc """
  Data directory. Contains `models.json`, `agents/`, `skills/`.

  Uses `XDG_DATA_HOME/code_puppy_ex` if set, otherwise falls back to `home_dir/`.
  """
  @spec data_dir() :: String.t()
  def data_dir do
    xdg_dir("XDG_DATA_HOME")
  end

  @doc """
  Cache directory. Contains autosaves and temporary data.

  Uses `XDG_CACHE_HOME/code_puppy_ex` if set, otherwise falls back to `home_dir/`.
  """
  @spec cache_dir() :: String.t()
  def cache_dir do
    xdg_dir("XDG_CACHE_HOME")
  end

  @doc """
  State directory. Contains command history and persistent runtime state.

  Uses `XDG_STATE_HOME/code_puppy_ex` if set, otherwise falls back to `home_dir/`.
  """
  @spec state_dir() :: String.t()
  def state_dir do
    xdg_dir("XDG_STATE_HOME")
  end

  # ── Config files ────────────────────────────────────────────────────────

  @doc "Path to the main config file `puppy.cfg`."
  @spec config_file() :: String.t()
  def config_file, do: Path.join(config_dir(), "puppy.cfg")

  @doc "Path to the MCP servers config file."
  @spec mcp_servers_file() :: String.t()
  def mcp_servers_file, do: Path.join(config_dir(), "mcp_servers.json")

  # ── Data files ──────────────────────────────────────────────────────────

  @doc "Path to the model registry JSON file."
  @spec models_file() :: String.t()
  def models_file, do: Path.join(data_dir(), "models.json")

  @doc "Path to user-added extra models file."
  @spec extra_models_file() :: String.t()
  def extra_models_file, do: Path.join(data_dir(), "extra_models.json")

  @doc "Path to the agents directory."
  @spec agents_dir() :: String.t()
  def agents_dir, do: Path.join(data_dir(), "agents")

  @doc "Path to the skills directory."
  @spec skills_dir() :: String.t()
  def skills_dir, do: Path.join(data_dir(), "skills")

  @doc "Path to the contexts directory."
  @spec contexts_dir() :: String.t()
  def contexts_dir, do: Path.join(data_dir(), "contexts")

  # ── Plugin / policy paths ──────────────────────────────────────────────

  @doc """
  Path to the model packs JSON file.

  Used by `ModelPacks` to persist user-defined model packs.
  """
  @spec model_packs_file() :: String.t()
  def model_packs_file, do: Path.join(data_dir(), "model_packs.json")

  @doc """
  Path to the user plugins directory.

  Used by `Plugins.Loader` for runtime plugin discovery.
  """
  @spec plugins_dir() :: String.t()
  def plugins_dir, do: Path.join(home_dir(), "plugins")

  @doc """
  Path to the user-level policy JSON file.

  Used by `PolicyConfig` and `PolicyEngine` for user-wide policy rules.
  """
  @spec user_policy_file() :: String.t()
  def user_policy_file, do: Path.join(config_dir(), "policy.json")

  @doc """
  Path to the project-local policy JSON file.

  Resolved relative to the current working directory (`.code_puppy/policy.json`).
  This is per-project config, NOT subject to the isolation guard.
  """
  @spec project_policy_file() :: String.t()
  def project_policy_file, do: Path.join([File.cwd!(), ".code_puppy", "policy.json"])

  @doc """
  Path to the Universal Constructor tools directory.

  Used by UC create_action, registry, and facade modules.
  """
  @spec universal_constructor_dir() :: String.t()
  def universal_constructor_dir, do: Path.join(plugins_dir(), "universal_constructor")

  @doc """
  Path to the ChatGPT models overlay JSON file.

  Used by `ModelRegistry` to load ChatGPT OAuth model definitions.
  """
  @spec chatgpt_models_file() :: String.t()
  def chatgpt_models_file, do: Path.join(data_dir(), "chatgpt_models.json")

  @doc """
  Path to the Claude models overlay JSON file.

  Used by `ModelRegistry` to load Claude Code OAuth model definitions.
  """
  @spec claude_models_file() :: String.t()
  def claude_models_file, do: Path.join(data_dir(), "claude_models.json")

  @doc """
  Path to the encrypted credentials store directory.

  Contains `store.json` — the AES-256-GCM encrypted credential store
  managed by `CodePuppyControl.Credentials`.
  """
  @spec credentials_dir() :: String.t()
  def credentials_dir, do: Path.join(data_dir(), "credentials")

  # ── Cache files ─────────────────────────────────────────────────────────

  @doc "Path to the autosaves directory."
  @spec autosave_dir() :: String.t()
  def autosave_dir, do: Path.join(cache_dir(), "autosaves")

  # ── State files ─────────────────────────────────────────────────────────

  @doc "Path to the command history file."
  @spec command_history_file() :: String.t()
  def command_history_file, do: Path.join(state_dir(), "command_history.txt")

  # ── Canonical Path Resolution ───────────────────────────────────────────

  @doc """
  Resolve a path to its canonical form by following all symlinks.

  Walks each path component, resolving symlinks at every level. Returns the
  fully resolved path. If max symlink depth (40) is exceeded, returns the
  last successfully resolved path. If the path or any component doesn't
  exist, returns the `Path.expand/1` result (symlinks can only be followed
  for existing filesystem entries).
  """
  @spec canonical_resolve(Path.t()) :: String.t()
  def canonical_resolve(path) do
    expanded = Path.expand(path)
    walk_components(Path.split(expanded), [], 0)
  end

  defp walk_components(remaining, acc, depth) when depth >= @max_symlink_depth do
    join_accumulated(acc, remaining)
  end

  defp walk_components([], acc, _depth) do
    Path.join(Enum.reverse(acc))
  end

  defp walk_components([component | rest], acc, depth) do
    current =
      case acc do
        [] -> component
        _ -> Path.join(Enum.reverse([component | acc]))
      end

    case :file.read_link(current) do
      {:ok, target} ->
        resolved = resolve_symlink_target(current, target)
        new_components = Path.split(resolved) ++ rest
        walk_components(new_components, [], depth + 1)

      {:error, _} ->
        walk_components(rest, [component | acc], depth)
    end
  end

  defp resolve_symlink_target(source, target) do
    if Path.type(target) == :absolute do
      Path.expand(target)
    else
      source
      |> Path.dirname()
      |> Path.join(target)
      |> Path.expand()
    end
  end

  defp join_accumulated(acc, remaining) do
    Path.join(Enum.reverse(acc) ++ remaining)
  end

  # ── Legacy Home Detection ───────────────────────────────────────────────

  @doc """
  Returns `true` if the given path resolves to a location under the legacy
  home directory (`~/.code_puppy`).

  Performs canonical resolution (following symlinks) before comparison.
  This blocks symlink attacks where a path like `/tmp/link → ~/.code_puppy/x`
  would bypass the isolation guard.
  """
  @spec in_legacy_home?(Path.t()) :: boolean()
  def in_legacy_home?(path) do
    resolved = canonical_resolve(path)
    legacy = legacy_home_dir()
    resolved == legacy or String.starts_with?(resolved, legacy <> "/")
  end

  # ── Utilities ───────────────────────────────────────────────────────────

  @doc """
  Ensure all standard directories exist with `0o700` permissions.
  Returns `:ok`.
  """
  @spec ensure_dirs!() :: :ok
  def ensure_dirs! do
    for dir <- [
          config_dir(),
          data_dir(),
          cache_dir(),
          state_dir(),
          skills_dir(),
          credentials_dir()
        ] do
      File.mkdir_p!(dir)
      # Best-effort permission set (no-op on some platforms)
      File.chmod(dir, 0o700)
    end

    :ok
  end

  @doc """
  Return a project-local agents directory if `.code_puppy/agents/` exists
  in the current working directory, or `nil`.
  """
  @spec project_agents_dir() :: String.t() | nil
  def project_agents_dir do
    path = Path.join(File.cwd!(), ".code_puppy/agents")
    if File.dir?(path), do: path, else: nil
  end

  # ── Private ─────────────────────────────────────────────────────────────

  defp xdg_dir(xdg_var) do
    cond do
      env = System.get_env("PUP_EX_HOME") ->
        env

      env = System.get_env("PUP_HOME") ->
        maybe_warn_deprecation("PUP_HOME")
        env

      true ->
        xdg_or_default(xdg_var)
    end
  end

  defp xdg_or_default(xdg_var) do
    case System.get_env(xdg_var) do
      nil ->
        case System.get_env("PUPPY_HOME") do
          nil ->
            Path.join(@home_dir, ".code_puppy_ex")

          home ->
            maybe_warn_deprecation("PUPPY_HOME")
            home
        end

      xdg_base ->
        Path.join(xdg_base, "code_puppy_ex")
    end
  end

  @doc false
  def maybe_warn_deprecation(env_var) do
    key = {:code_puppy_control, :deprecation_warned, env_var}

    unless :persistent_term.get(key, false) do
      :persistent_term.put(key, true)

      Logger.warning("#{env_var} is deprecated for Elixir pup-ex. Use PUP_EX_HOME instead.")
    end

    :ok
  end
end
